Update UMAT interface vector order to typical use

recuero opened the pull request:
Closes #20111

recuero committed:
Update UMAT interface vector order to typical use (#20111)

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/938832/) on 6391693 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/20112/site)

This comment will be updated on new commits.

dschwen commented:
How is this not diffing more things?!

recuero commented:
Because we don't have real life tests!!!

dschwen commented:
Let's make sure to change that in this pr then. Will require an anisotropic elasticity tensor

dschwen commented on modules/tensor_mechanics/src/materials/abaqus/AbaqusUMATStress.C:
```diff
@@ -170,8 +170,10 @@ AbaqusUMATStress::computeQpStress()
+  // (commercial software)
   static const std::array<std::pair<unsigned int, unsigned int>, 6> component{
-      {{0, 0}, {1, 1}, {2, 2}, {1, 2}, {0, 2}, {0, 1}}};
+      {{0, 0}, {1, 1}, {2, 2}, {0, 1}, {0, 2}, {1, 2}}};
```
By "typical use" you mean not standard Voigt notation order?

recuero replied:
I was trying to refer to the use of the UMAT interface. I guess there can be a variety of not standard Voigt order.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/938842/) on 6391693 wanted to post the following:


## Framework coverage

Coverage did not change


## Modules coverage

### Tensor mechanics

<table>
  <thead>
    <tr>
      <th></th>
      <th>e6f1df</th>
      <th colspan=4>#20112 639169</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>89.21%</td>
      <td>89.21%</td>
      <td>-</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>25223</td>
      <td>25223</td>
      <td>-</td>
      <td>1</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>3050</td>
      <td>3050</td>
      <td>-</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/20112/coverage_diff/tensor_mechanics.html)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/20112/coverage/tensor_mechanics)

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/20112/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/20112/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/20112/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/20112/coverage/contact)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/20112/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/20112/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/20112/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/20112/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/20112/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/20112/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/20112/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/20112/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/20112/coverage/navier_stokes)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/20112/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/20112/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/20112/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/20112/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/20112/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/20112/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/20112/coverage/richards)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/20112/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/20112/coverage/tensor_mechanics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/20112/coverage/xfem)

</details>



This comment will be updated on new commits.

recuero commented:
> Let's make sure to change that in this pr then. Will require an anisotropic elasticity tensor

For the models that we are testing, it may require an anisotropic elasticity tensor. But we would still be cooking up a problem for this particular issue (I was not referring to this, but to the lack of more advanced UMAT routines). Do you want to have a test with some elastic anisotropy added to this PR?

recuero committed:
Add tests that check for shear term order for UMAT interface (#20111)

This is done here in two ways:

 - First comparison is performed for shear stresses when a single element
   is sheared through UMAT and through a MOOSE stress calculator when
   shear stiffness is anisotropic, i.e. XY and XZ have a different
   "G". Results match for low strain (for large strains rotations in the UMAT
   routine need to be performed)
 - A Jacobian tester test is added where the UMAT Jacobian is on the order of
   MOOSE's (actually, somehow slightly better). With this change, the UMAT test
   Jacobian relative tolerance goes from ~20% to ~4%.

dschwen commented on modules/tensor_mechanics/src/materials/abaqus/AbaqusUMATStress.C:
```diff
@@ -280,34 +282,32 @@ AbaqusUMATStress::computeQpStress()
-      _aqDDSDDE[35]  // C1212
-  }});
+  // Build Jacobian matrix from UMAT's Voigt non-standard order to fourth order tensor.
+  _jacobian_mult[_qp].zero();
```
remove
```suggestion
```

recuero replied:
Just did.

dschwen reviewed (approved):
Looks good to me
